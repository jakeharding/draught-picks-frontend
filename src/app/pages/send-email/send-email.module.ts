import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { IonicModule } from '@ionic/angular';
import { SendEmailPage } from './send-email.page';
import { ComponentsModule } from '../../components/components.module';


@NgModule({
  imports: [
    CommonModule,
    IonicModule,
    ComponentsModule,
    ReactiveFormsModule,
    RouterModule.forChild([{ path: '', component: SendEmailPage}])
  ],
  declarations: [SendEmailPage]
})
export class SendEmailPageModule {}
