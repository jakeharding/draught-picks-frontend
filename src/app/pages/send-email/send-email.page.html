<!--
  Generated template for the ResendEmailPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content>
  <ion-row>
    <ion-col class="ion-text-center"
             size="12" sizeSm="10" offsetSm="1" sizeMd="6" offsetMd="3" sizeLg="6" offsetLg="3" sizeXl="4" offsetXl="4">
        <h1>Send {{ reason?.title }} Email</h1>
        <br>
        <div class="ion-text-left">
          <div *ngIf="reason">
            <form [formGroup]="emailForm" (submit)="sendEmail()">
              <p>An email with instructions on how to {{ reason?.message }} will be sent if the address is on record.</p>
              <p>Please enter the email address associated with your account:</p>
              <ion-item>
                <ion-label class="ion-text-left" position="floating">Email</ion-label>
                <ion-input type="text" formControlName="email"></ion-input>
              </ion-item>
              <button type="submit" class="ion-page-hidden"></button>
              <ion-button
                  (click)="reason.sendEmail()"
                  class="ion-margin-top"
                  [disabled]="emailForm.invalid">
                Submit
              </ion-button>
            </form>
          </div>
          <div *ngIf="!reason">
            <p>Please select the reason for sending an email below:</p>
            <p><a class="pointer" routerLink="/send-email/confirm" routerDirection="forward">Resend Email Confirmation</a></p>
            <p><a class="pointer" routerLink="/send-email/reset" routerDirection="forward">Reset password</a></p>
          </div>
        </div>

    </ion-col>
  </ion-row>

</ion-content>
