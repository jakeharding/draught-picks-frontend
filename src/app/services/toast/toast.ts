import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class ToastProvider {
  static readonly defaultErrorMsg = 'We are having trouble connecting at the moment. ' +
  'Please check your network connection and try again.';
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string = ToastProvider.defaultErrorMsg): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  async createToast(message: string,
                    cssClass: string,
                    duration: number = 3000,
                    position: 'top' | 'bottom' | 'middle' = 'top'
              ) {
    const toast = await this.toastCtrl.create({ message, cssClass, duration, position });
    await toast.present();
  }

}
