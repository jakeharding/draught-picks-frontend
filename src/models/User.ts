/**
 * User.ts
 *
 * Created by jake
 * Created on 2/27/18
 *
 * Model user attributes
 */
import Beer from "./Beer";


export default interface User {
  first_name: string;
  last_name: string;
  uuid: string;
  username: string;
  password: string;
  date_of_birth: Date;
  email: string;
  weight: number;
  favorite_beers: Array<Beer>;
  rated_beers: Array<Beer>;
  recent_beers: Array<Beer>;
}
