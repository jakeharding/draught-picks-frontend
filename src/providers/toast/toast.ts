import { Injectable } from '@angular/core';
import { ToastController } from 'ionic-angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable()
export class ToastProvider {
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  createToast(message: string,
              cssClass:string,
              duration: number = 3000,
              position: string = 'top'
              ) {
    this.toastCtrl.create({ message, cssClass, duration, position }).present();
  }

}
