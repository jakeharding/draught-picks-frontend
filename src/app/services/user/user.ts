import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import PageResponse from '../../models/PageResponse';
import User from '../../models/User';
import AuthResponse from '../../models/AuthResponse';
import BeerRating from '../../models/BeerRating';
import Beer from '../../models/Beer';

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 */
@Injectable({
  providedIn: 'root',
})
export class UserProvider {
  private readonly url: string;
  private readonly resendConfirmEmailUrl: string;
  private readonly confirmEmailUrl: string;
  private readonly passwordResetUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
    this.confirmEmailUrl = `${this.url}/confirm-email`;
    this.passwordResetUrl = `${this.url}/reset-email`;
  }

  create(userData: AuthResponse | Beer | BeerRating | User) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }

  confirmEmail({key}): Promise<any> {
    return this.http.put(this.confirmEmailUrl, {confirm_key: key}).toPromise();
  }

  sendPasswordResetEmail(emailAddress: string): Promise<any> {
    return this.http.post(this.passwordResetUrl, {email: emailAddress}).toPromise();
  }
}
