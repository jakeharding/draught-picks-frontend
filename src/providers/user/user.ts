import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import Env from "../../env";
import PageResponse from "../../models/PageResponse";
import User from "../../models/User";

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
**/
@Injectable()
export class UserProvider {
  url: string;
  resendConfirmEmailUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
  }

  create(userData: any) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<Object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }
}
