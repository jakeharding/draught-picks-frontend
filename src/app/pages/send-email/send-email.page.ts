import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { NavController } from '@ionic/angular';
import { ToastProvider } from '../../services/toast/toast';
import { UserProvider } from '../../services/user/user';
import { BasePage } from '../BasePage';
import { ActivatedRoute } from '@angular/router';

export enum EmailReason {
  RESET = 'reset',
  CONFIRM = 'confirm'
}

interface Reason {
  message: string;
  title: string;
  sendEmail: () => void;
}

@Component({
  selector: 'app-send-email',
  templateUrl: './send-email.page.html',
  styleUrls: ['./send-email.page.scss'],
})
export class SendEmailPage extends BasePage {

  REASON_META: any;
  emailForm: FormGroup;
  sendEmail: () => void;
  reason: Reason;

  constructor(public formBuilder: FormBuilder,
              public navCtrl: NavController,
              private toastProvider: ToastProvider,
              private route: ActivatedRoute,
              private userProvider: UserProvider) {
    super(`send-email/${route.snapshot.paramMap.get('reason')}`);
    this.REASON_META = {
      [EmailReason.CONFIRM]: {
        message: 'confirm your email.',
        title: 'Confirm',
        sendEmail: () => this.sendConfirmEmail()
      },
      [EmailReason.RESET]: {
        message: 'reset your password.',
        title: 'Password Reset',
        sendEmail: () => this.sendResetEmail()
      }
    };
    this.reason = this.REASON_META[route.snapshot.paramMap.get('reason')];

    this.emailForm = this.formBuilder.group({
      email: ['', [Validators.required, Validators.email]]
    });
  }

  sendResetEmail() {
    return this.userProvider.sendPasswordResetEmail(this.emailForm.value.email).then(r =>  {
      this.toastProvider.successToast('The email will arrive soon.');
    }).catch(err => {
      console.error(err);
      this.toastProvider.errorToast('The email cannot be sent at this time. Please check the email address and try again');
    });
  }

  setReason(reason: EmailReason) {
    this.reason = this.REASON_META[reason];
  }

  sendConfirmEmail() {
    this.userProvider.resendConfirmEmail(this.emailForm.value).then(() => {
      this.toastProvider.successToast(
        'Email has been sent. Please check you spam folder and visit this page again if you don\'t receive it.');
    }, () => {
      this.toastProvider.errorToast('Unable to send the email. Please check your connection and try again.');
    });
  }

}
