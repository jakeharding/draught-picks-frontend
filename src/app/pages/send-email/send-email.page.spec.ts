import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { EmailReason, SendEmailPage } from './send-email.page';
import { FormBuilder, FormGroupDirective } from '@angular/forms';
import { NavController, NavParams } from '@ionic/angular';
import { ToastProvider } from '../../services/toast/toast';
import { UserProvider } from '../../services/user/user';
import { ActivatedRoute } from '@angular/router';
import { mockActivatedRoute } from '../../../../setup-jest';
import { By } from '@angular/platform-browser';
import User from '../../models/User';

describe('SendEmailPage', () => {
  let component: SendEmailPage;
  let fixture: ComponentFixture<SendEmailPage>;

  const mockUserProvider = {
    resendConfirmEmail: jest.fn(() => Promise.resolve()),
    sendPasswordResetEmail: jest.fn()
  };

  const mockToastProvider = {
    successToast: jest.fn(),
    errorToast: jest.fn()
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [SendEmailPage, FormGroupDirective],
      providers: [
        { provide: NavController, useValue: {}},
        { provide: NavParams, useValue: {}},
        { provide: ToastProvider, useValue: mockToastProvider},
        { provide: UserProvider, useValue: mockUserProvider},
        { provide: ActivatedRoute, useValue: mockActivatedRoute},
        FormBuilder
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SendEmailPage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the SendEmailPage with reason undefined', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
    fixture.detectChanges();
    expect(fixture.debugElement.queryAll(By.css('a.pointer'))).toHaveLength(2);
  });

  test('setReason', () => {
    component.setReason(EmailReason.CONFIRM);
    expect(component.reason).toBeDefined();
    expect(component.reason.message).toBe('confirm your email.');
  });

  test('sendConfirmEmail successToast is called on promise resolve', async () => {
    await component.sendConfirmEmail();
    expect(mockUserProvider.resendConfirmEmail).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.successToast).toHaveBeenCalledTimes(1);
  });

  test('sendConfirmEmail successToast is called on promise reject', async () => {
    mockUserProvider.resendConfirmEmail.mockReturnValue(Promise.reject('error'));
    await component.sendConfirmEmail();
    expect(mockUserProvider.resendConfirmEmail).toHaveBeenCalledTimes(1);
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
  });

  test('sendResetEmail calls userProvider.sendPasswordResetEmail', async () => {
    mockUserProvider.sendPasswordResetEmail.mockResolvedValue('');
    component.emailForm.value.email = 't@t.com';
    await component.sendResetEmail();
    expect(mockUserProvider.sendPasswordResetEmail).toHaveBeenCalledTimes(1);
    expect(mockUserProvider.sendPasswordResetEmail).toHaveBeenCalledWith('t@t.com');
    expect(mockToastProvider.successToast).toHaveBeenCalledTimes(1);
  });

  test('sendResetEmail calls errorToast when not successful', async () => {
    mockUserProvider.sendPasswordResetEmail.mockRejectedValue('BOOM');
    component.emailForm.value.email = 't@t.com';
    await component.sendResetEmail();
    expect(mockUserProvider.sendPasswordResetEmail).toHaveBeenCalledTimes(1);
    expect(mockUserProvider.sendPasswordResetEmail).toHaveBeenCalledWith('t@t.com');
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
  });
});
