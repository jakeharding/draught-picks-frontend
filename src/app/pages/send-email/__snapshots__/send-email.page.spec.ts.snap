// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendEmailPage should initialize the SendEmailPage 1`] = `
<app-send-email
  emailForm={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  navCtrl={[Function Object]}
  toastProvider={[Function Object]}
  userProvider={[Function Object]}
>
  <ion-header>
    <draught-header />
  </ion-header><ion-content>
    <ion-row>
      <ion-col
        class="ion-text-center"
        offsetlg="3"
        offsetmd="3"
        offsetsm="1"
        offsetxl="4"
        size="12"
        sizelg="6"
        sizemd="6"
        sizesm="10"
        sizexl="4"
      >
        <form>
          <h1>
            Send Email
          </h1>
          <br />
          <p>
            Please enter the email address below:
          </p>
          <ion-item>
            <ion-label
              class="ion-text-left"
              position="floating"
            >
              Email
            </ion-label>
            <ion-input
              formcontrolname="email"
              type="text"
            />
          </ion-item>
        </form>
        <ion-button
          margin-top="20px"
        >
           Submit 
        </ion-button>
      </ion-col>
    </ion-row>
  </ion-content>
</app-send-email>
`;
